rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -734,19 +698,24 @@ def train(
         else:
             self._train_with_independent_vocab(attribute_texts)
 
-        # transform for all attributes
-        for attribute in self._attributes:
-            sequence_features, sentence_features = self._get_featurized_attribute(
-                attribute, processed_attribute_tokens[attribute]
-            )
+    def train(
+        self,
+        training_data: TrainingData,
+        config: Optional[RasaNLUModelConfig] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Applies to featurizer to the training data."""
 
-            if sequence_features and sentence_features:
-                self._set_attribute_features(
-                    attribute,
-                    sequence_features,
-                    sentence_features,
-                    training_data.training_examples,
-                )
+        self.initialize(training_data, config, **kwargs)
+
+        msg_pipe = training_data.add_features(None, "test2")
+        try:
+            msg = next(msg_pipe)
+            while msg:
+                features = self.msg_to_features(msg)
+                msg = msg_pipe.send(features)
```

Comment:
using `send` is neat!

Reply:
Yeah first time for me! It's cool, but easy to get wrong. At first I only stored every second feature set because I both consumed the generator and used send too. Didn't know that send would also make the generator continue and return the next yield immediately.

Reply:
It makes sense once you think about it but it's not intuitive 🙈

Reply:
I still haven't been able to wrap my head around how it's working, but I okay considering it as some sort of magic 😅  for now which sets the features computed by the specific featurizer for that message, right?

Reply:
>  for now which sets the features computed by the specific featurizer for that message, right?

Yes. With the advantage that we control how features are added outside of the components (which in turn means the `Component`s don't have to know about these details). Whether we write the features to disks for chunks or keep them in memory is completely abstracted and exchangable.

Reply:
And `Components` are also not aware where data is coming from. It could be in-memory, read from disk or some stuff is coming from in-memory and some is coming from disk.

Reply:
> With the advantage that we control how features are added outside of the components (which in turn means the Components don't have to know about these details).

But we do make a distinction between different kind of features? for e.g. - sparse and dense features.